In [13]:
# 라이브러리 임포트
import os
import numpy as np
import pandas as pd
import chardet

# 사용자 정의함수 경로 설정
import sys
sys.path.append("G:/내 드라이브/Source/_Custom_Function")
# 사용자 정의함수 불러오기

import time

import ExcelFile_Merge as em
import ExcelFile_Read as er
import ExcelFile_EDA as ee

In [56]:
Default_path = 'G:\\내 드라이브\\DataSet\\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\\23년\\'

---
### 경로 파일 체크
- 23.01-23.03 (17일까지만 있음)

In [58]:
Sub_path = '23.01-23.03 (17일까지만 있음)\\'

In [60]:
er.Columns_Check(Default_path + Sub_path) 

file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.01-23.03 (17일까지만 있음)\
0  file :  23.01.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
1  file :  23.02.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
2  file :  23.03.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
O,O,O,O


In [62]:
_files = ['23.01.csv', '23.02.csv', '23.03.csv']
for _file in _files:
    df = em.Read_File(Default_path + Sub_path + _file)
    # '일시', '대여소번호', '대여소명', '시간대' 컬럼의 중복값 확인
    duplicates = df.duplicated(subset=['일시', '대여소번호', '대여소명', '시간대'], keep=False)
    # 중복된 행 출력
    duplicate_rows = df[duplicates]
    # 결과 출력
    print(duplicate_rows.shape[0])

read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.01-23.03 (17일까지만 있음)\23.01.csv , encoding :  EUC-KR
0
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.01-23.03 (17일까지만 있음)\23.02.csv , encoding :  EUC-KR
0
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.01-23.03 (17일까지만 있음)\23.03.csv , encoding :  EUC-KR
0


---
### 경로 파일 체크
- 23.04-23.12_컬럼명추가

In [64]:
Sub_path = '23.04-23.12_컬럼명추가\\'

In [66]:
er.Columns_Check(Default_path + Sub_path) 

file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.04-23.12_컬럼명추가\
0  file :  data_230401~230415_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
1  file :  data_230416~230430_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
2  file :  data_230501~230515_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
3  file :  data_230516~230531_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
4  file :  data_230601~230615_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
5  file :  data_230616~230630_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
6  file :  data_230701~230715_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
7  file :  data_230716~230731_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
8  file :  data_230801~230815_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
9  file :  data_230816~230831_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
10  file :  data_230901~230915_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
11  file :  data_230916~230930_컬럼추가.csv
columns :  일시,대여소번호,대여소명,시간대,거치대수량
12  file :  data_231001

In [68]:
# 파일 경로명 변경
file_list = os.listdir(Default_path + Sub_path)
file_lists = [file for file in file_list if file.endswith((".csv", ".xlsx"))]
file_lists.sort()
print('file_lists : \n', '\n '.join(file_lists))
print('===========================================')

for _, filename in enumerate(file_lists):
    try:
        print('===================================')
        df = em.Read_File(Default_path + Sub_path + filename)
        # '일시', '대여소번호', '대여소명', '시간대' 컬럼의 중복값 확인
        duplicates = df.duplicated(subset=['일시', '대여소번호', '대여소명', '시간대'], keep=False)
        # 중복된 행 출력
        duplicate_rows = df[duplicates]
        # 결과 출력
        print(f' - [중복] : {duplicate_rows.shape[0]} / {df.shape[0]}')
    except:
        print('- Err : ', savefile)     

file_lists : 
 data_230401~230415_컬럼추가.csv
 data_230416~230430_컬럼추가.csv
 data_230501~230515_컬럼추가.csv
 data_230516~230531_컬럼추가.csv
 data_230601~230615_컬럼추가.csv
 data_230616~230630_컬럼추가.csv
 data_230701~230715_컬럼추가.csv
 data_230716~230731_컬럼추가.csv
 data_230801~230815_컬럼추가.csv
 data_230816~230831_컬럼추가.csv
 data_230901~230915_컬럼추가.csv
 data_230916~230930_컬럼추가.csv
 data_231001~231015_컬럼추가.csv
 data_231016~231031_컬럼추가.csv
 data_231101~231115_컬럼추가.csv
 data_231116~231130_컬럼추가.csv
 data_231201~231215_컬럼추가.csv
 data_231216~231231_컬럼추가.csv
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.04-23.12_컬럼명추가\data_230401~230415_컬럼추가.csv , encoding :  EUC-KR
 - [중복] : 5762759 / 5762760
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.04-23.12_컬럼명추가\data_230416~230430_컬럼추가.csv , encoding :  EUC-KR
 - [중복] : 5767200 / 5767200
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.04-23.12_컬럼명추가\data_230501~2305

data_230401~230415_컬럼추가.csv
 data_230416~230430_컬럼추가.csv
 data_230501~230515_컬럼추가.cs
 data_230516~230531_컬럼추가.cv
 data_230601~230615_컬럼추가.sv
 data_230616~230630_컬럼추가csv
 data_230701~230715_컬럼추.csv
 data_230716~230731_컬럼가.csv
 data_230801~230815_컬추가.csv
 data_230816~230831_럼추가.csv
 data_230901~230915컬럼추가.csv
 data_230916~23093_컬럼추가.csv
 data_231001~23105_컬럼추가.csv
 data_231016~23131_컬럼추가.csv
 data_231101~23115_컬럼추가.csv
 data_231116~21130_컬럼추가.csv
 data_231201~31215_컬럼추가.csv
 data_231216~231231_컬럼추가.csv

In [72]:
filename = 'data_230516~230531_컬럼추가.csv'
df = em.Read_File(Default_path + Sub_path + filename)
grb_df = df.groupby(['일시', '대여소번호', '대여소명', '시간대'])['거치대수량'].count()
grb_df = grb_df.reset_index()

print(df['거치대수량'].value_counts())
print(f'df.shape[0] : {df.shape[0]}')

print(grb_df['거치대수량'].value_counts())
print(f'grb_df.shape[0] : {grb_df.shape[0]}')

read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23.04-23.12_컬럼명추가\data_230516~230531_컬럼추가.csv , encoding :  EUC-KR
거치대수량
0      1045745
1       620700
2       456045
3       357339
4       298315
        ...   
224          1
223          1
230          1
233          1
220          1
Name: count, Length: 229, dtype: int64
df.shape[0] : 6162491
거치대수량
6    1027079
5          2
3          1
4          1
Name: count, dtype: int64
grb_df.shape[0] : 1027083


In [66]:
df[df['거치대수량'] == 0].head(5)

일시  대여소번호                대여소명  시간대  거치대수량
5   2023-04-16    107     107. 신한은행 서교동지점    0      0
8   2023-04-16    113   113. 홍대입구역 2번출구 앞    0      0
9   2023-04-16    114   114. 홍대입구역 8번출구 앞    0      0
26  2023-04-16    134        134. 연세로 명물길    0      0
29  2023-04-16    138  138. 신촌동 제1공영주차장 앞    0      0

In [68]:
cond1 = df['일시'] == '2023-04-16'
cond2 = df['대여소번호'] == 107
cond3 = df['시간대'] == 0
df[cond1 & cond2 & cond3]

일시  대여소번호             대여소명  시간대  거치대수량
5      2023-04-16    107  107. 신한은행 서교동지점    0      0
2675   2023-04-16    107  107. 신한은행 서교동지점    0      0
5345   2023-04-16    107  107. 신한은행 서교동지점    0      0
8015   2023-04-16    107  107. 신한은행 서교동지점    0      1
10685  2023-04-16    107  107. 신한은행 서교동지점    0      1
13355  2023-04-16    107  107. 신한은행 서교동지점    0      0

In [43]:
grb_df[grb_df['거치대수량'].isin([1, 5])]

일시  대여소번호              대여소명  시간대  거치대수량
697185  2023-04-11   4546  4546. 목동운동장 주경기장   14      1
702979  2023-04-11   5095      5095. 원당근린공원   14      5

In [39]:
cond1 = df['일시'] == '2023-04-01'
cond2 = df['대여소번호'] == 102
cond3 = df['시간대'] == 1
df[cond1 & cond2 & cond3]

일시  대여소번호             대여소명  시간대  거치대수량
16002  2023-04-01    102  102. 망원역 1번출구 앞    1     38
18669  2023-04-01    102  102. 망원역 1번출구 앞    1     38
21336  2023-04-01    102  102. 망원역 1번출구 앞    1     36
24003  2023-04-01    102  102. 망원역 1번출구 앞    1     36
26670  2023-04-01    102  102. 망원역 1번출구 앞    1     36
29337  2023-04-01    102  102. 망원역 1번출구 앞    1     36

In [45]:
cond1 = df['일시'] == '2023-04-01'
cond2 = df['대여소번호'] == 102
cond3 = df['시간대'] == 2
df[cond1 & cond2 & cond3]

일시  대여소번호             대여소명  시간대  거치대수량
32004  2023-04-01    102  102. 망원역 1번출구 앞    2     36
34671  2023-04-01    102  102. 망원역 1번출구 앞    2     35
37338  2023-04-01    102  102. 망원역 1번출구 앞    2     34
40005  2023-04-01    102  102. 망원역 1번출구 앞    2     34
42672  2023-04-01    102  102. 망원역 1번출구 앞    2     34
45339  2023-04-01    102  102. 망원역 1번출구 앞    2     34

---
## 평균 값으로 저장

In [91]:
filename = 'data_230516~230531_컬럼추가.csv'
df = em.Read_File(Default_path + Sub_path + filename)
grb_df = df.groupby(['일시', '대여소번호', '대여소명', '시간대'])['거치대수량'].mean()
grb_df = np.floor(grb_df).astype(int)
grb_df = grb_df.reset_index()

print(df['거치대수량'].value_counts())
print(f'df.shape[0] : {df.shape[0]}')

print(grb_df['거치대수량'].value_counts())
print(f'grb_df.shape[0] : {grb_df.shape[0]}')

read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23.04-23.12_컬럼명추가\data_230516~230531_컬럼추가.csv , encoding :  EUC-KR
거치대수량
0      1045745
1       620700
2       456045
3       357339
4       298315
        ...   
224          1
223          1
230          1
233          1
220          1
Name: count, Length: 229, dtype: int64
df.shape[0] : 6162491
거치대수량
0      197610
1      101377
2       72632
3       57240
4       47689
        ...  
164         1
171         1
160         1
165         1
232         1
Name: count, Length: 202, dtype: int64
grb_df.shape[0] : 1027083


In [92]:
grb_df.head(3)

일시  대여소번호             대여소명  시간대  거치대수량
0  2023-05-16    102  102. 망원역 1번출구 앞    0      8
1  2023-05-16    102  102. 망원역 1번출구 앞    1      7
2  2023-05-16    102  102. 망원역 1번출구 앞    2      6

In [87]:
cond1 = df['일시'] == '2023-04-16'
cond2 = df['대여소번호'] == 107
cond3 = df['시간대'] == 0
df[cond1 & cond2 & cond3]

Empty DataFrame
Columns: [일시, 대여소번호, 대여소명, 시간대, 거치대수량]
Index: []

---
## 병합
- 23.04-23.12_컬럼명추가

In [97]:
df_sum = pd.DataFrame()

In [99]:
Sub_path = '23.04-23.12_컬럼명추가\\'

In [101]:
# 파일 경로명 변경
file_list = os.listdir(Default_path + Sub_path)
file_lists = [file for file in file_list if file.endswith((".csv", ".xlsx"))]
file_lists.sort()
print('file_lists : \n', '\n '.join(file_lists))
print('===========================================')

for _, filename in enumerate(file_lists):
    print('===================================')
    df = em.Read_File(Default_path + Sub_path + filename)
    
    grb_df = df.groupby(['일시', '대여소번호', '대여소명', '시간대'])['거치대수량'].mean()
    grb_df = np.floor(grb_df).astype(int)
    grb_df = grb_df.reset_index()

    df_sum = pd.concat([grb_df, df_sum], axis = 0)
    del df  # 메모리 정리
    del grb_df  # 메모리 정리

# for문 종료 후, 실행
df_sum.to_parquet(Default_path + Sub_path + 'sum_df' + '.parquet', index=False)
del df_sum

file_lists : 
 data_230401~230415_컬럼추가.csv
 data_230416~230430_컬럼추가.csv
 data_230501~230515_컬럼추가.csv
 data_230516~230531_컬럼추가.csv
 data_230601~230615_컬럼추가.csv
 data_230616~230630_컬럼추가.csv
 data_230701~230715_컬럼추가.csv
 data_230716~230731_컬럼추가.csv
 data_230801~230815_컬럼추가.csv
 data_230816~230831_컬럼추가.csv
 data_230901~230915_컬럼추가.csv
 data_230916~230930_컬럼추가.csv
 data_231001~231015_컬럼추가.csv
 data_231016~231031_컬럼추가.csv
 data_231101~231115_컬럼추가.csv
 data_231116~231130_컬럼추가.csv
 data_231201~231215_컬럼추가.csv
 data_231216~231231_컬럼추가.csv
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23.04-23.12_컬럼명추가\data_230401~230415_컬럼추가.csv , encoding :  EUC-KR
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23.04-23.12_컬럼명추가\data_230416~230430_컬럼추가.csv , encoding :  EUC-KR
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23.04-23.12_컬럼명추가\data_230501~230515_컬럼추가.csv , encoding :  EUC-KR
read_file - file_path :  G:\내 드라이브\

In [105]:
df_2 = em.Read_File(Default_path + Sub_path + 'sum_df.parquet')

In [107]:
df_2.shape[0]

16968692

In [109]:
df_2.head()

일시  대여소번호             대여소명  시간대  거치대수량
0  2023-12-16    102  102. 망원역 1번출구 앞    0     34
1  2023-12-16    102  102. 망원역 1번출구 앞    1     34
2  2023-12-16    102  102. 망원역 1번출구 앞    2     33
3  2023-12-16    102  102. 망원역 1번출구 앞    3     34
4  2023-12-16    102  102. 망원역 1번출구 앞    4     32

---
### 중복확인

In [113]:
# '일시', '대여소번호', '대여소명', '시간대' 컬럼의 중복값 확인
duplicates = df_2.duplicated(subset=['일시', '대여소번호', '대여소명', '시간대'], keep=False)
# 중복된 행 출력
duplicate_rows = df_2[duplicates]
# 결과 출력
print(f' - [중복] : {duplicate_rows.shape[0]} / {df_2.shape[0]}')

 - [중복] : 0 / 16968692


---
## 병합
- 23.01-23.03 (17일까지만 있음)

In [70]:
df_sum = pd.DataFrame()

In [72]:
Sub_path = '23.01-23.03 (17일까지만 있음)\\'

In [74]:
# 파일 경로명 변경
file_list = os.listdir(Default_path + Sub_path)
file_lists = [file for file in file_list if file.endswith((".csv", ".xlsx"))]
file_lists.sort()
print('file_lists : \n', '\n '.join(file_lists))
print('===========================================')

for _, filename in enumerate(file_lists):
    print('===================================')
    df = em.Read_File(Default_path + Sub_path + filename)
    
    df_sum = pd.concat([df, df_sum], axis = 0)
    del df  # 메모리 정리    

# for문 종료 후, 실행
df_sum.to_parquet(Default_path + Sub_path + '23.01~03(병합)' + '.parquet', index=False)
del df_sum

file_lists : 
 23.01.csv
 23.02.csv
 23.03.csv
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.01-23.03 (17일까지만 있음)\23.01.csv , encoding :  EUC-KR
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.01-23.03 (17일까지만 있음)\23.02.csv , encoding :  EUC-KR
read_file - file_path :  G:\내 드라이브\DataSet\서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)\23년\23.01-23.03 (17일까지만 있음)\23.03.csv , encoding :  EUC-KR


---
## 최종병합

In [84]:
df_sum = pd.DataFrame()

In [78]:
df1 = em.Read_File(Default_path + '23.01-23.03 (17일까지만 있음)\\23.01~03(병합).parquet')
df2 = em.Read_File(Default_path + '23.04-23.12_컬럼명추가\\data_230401~231231(중복제거).parquet')

In [79]:
df1.head(2)

일시  대여소번호             대여소명  시간대  거치대수량
0  2023-03-01    102  102. 망원역 1번출구 앞    0     22
1  2023-03-01    102  102. 망원역 1번출구 앞    1     19

In [46]:
df2.head(2)

일시  대여소번호             대여소명  시간대  거치대수량
0  2023-12-16    102  102. 망원역 1번출구 앞    0     34
1  2023-12-16    102  102. 망원역 1번출구 앞    1     34

In [48]:
df_sum = pd.concat([df1, df2], axis = 0)

In [50]:
df_sum.head(2)

일시  대여소번호             대여소명  시간대  거치대수량
0  2023-01-01    102  102. 망원역 1번출구 앞    0     15
1  2023-01-01    102  102. 망원역 1번출구 앞    1     14

In [82]:
cond1 = (df1['일시'] == '2023-01-01')
cond2 = (df1['대여소번호'] == 102)
cond3 = (df1['대여소명'] == '102. 망원역 1번출구 앞')
cond4 = (df1['시간대'] == 0)
df1[cond1 & cond2 & cond3 & cond4]

일시  대여소번호             대여소명  시간대  거치대수량
2097150  2023-01-01    102  102. 망원역 1번출구 앞    0     15

In [54]:
cond1 = (df2['일시'] == '2023-01-01')
cond2 = (df2['대여소번호'] == 102)
cond3 = (df2['대여소명'] == '102. 망원역 1번출구 앞')
cond4 = (df2['시간대'] == 0)
df2[cond1 & cond2 & cond3 & cond4]

Empty DataFrame
Columns: [일시, 대여소번호, 대여소명, 시간대, 거치대수량]
Index: []

In [34]:
cond1 = (df_sum['일시'] == '2023-01-01')
cond2 = (df_sum['대여소번호'] == 102)
cond3 = (df_sum['대여소명'] == '102. 망원역 1번출구 앞')
cond4 = (df_sum['시간대'] == 0)
df_sum[cond1 & cond2 & cond3 & cond4]

일시  대여소번호             대여소명  시간대  거치대수량
0        2023-01-01    102  102. 망원역 1번출구 앞    0     15
5242873  2023-01-01    102  102. 망원역 1번출구 앞    0     15

In [88]:
df1 = em.Read_File(Default_path + '23.01-23.03 (17일까지만 있음)\\23.01~03(병합).parquet')
df2 = em.Read_File(Default_path + '23.04-23.12_컬럼명추가\\data_230401~231231(중복제거).parquet')
    
df_sum = pd.concat([df1, df2], axis = 0)
del df1  # 메모리 정리    
del df2  # 메모리 정리    

# for문 종료 후, 실행
df_sum.to_parquet(Default_path + '서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)_2023.parquet', index=False)
del df_sum

In [90]:
df_2 = em.Read_File(Default_path + '서울특별시 대여소별 공공자전거 대여가능 수량(1시간 단위)_2023.parquet')

In [91]:
df_2.shape[0]

20114417

In [94]:
cond1 = (df_2['일시'] == '2023-01-01')
cond2 = (df_2['대여소번호'] == 102)
cond3 = (df_2['대여소명'] == '102. 망원역 1번출구 앞')
cond4 = (df_2['시간대'] == 0)
df_2[cond1 & cond2 & cond3 & cond4]

일시  대여소번호             대여소명  시간대  거치대수량
2097150  2023-01-01    102  102. 망원역 1번출구 앞    0     15